File: leetcode-04/solutions/cpp/0695.cpp
class Solution {
 public:
  int maxAreaOfIsland(vector<vector<int>>& grid) {
    if (grid.empty()) return 0;

    const int m = grid.size();
    const int n = grid[0].size();

    int ans = 0;
    vector<vector<bool>> visited(m, vector<bool>(n));

    for (int i = 0; i < m; i++)
      for (int j = 0; j < n; j++) ans = max(ans, dfs(grid, i, j, visited));

    return ans;
  }

 private:
  int dfs(vector<vector<int>>& grid, int i, int j,
          vector<vector<bool>>& visited) {
    if (i < 0 || j < 0 || i >= grid.size() || j >= grid[0].size() ||
        visited[i][j] || grid[i][j] == 0)
      return 0;

    visited[i][j] = true;

    return dfs(grid, i + 1, j, visited) + dfs(grid, i - 1, j, visited) +
           dfs(grid, i, j + 1, visited) + dfs(grid, i, j - 1, visited) + 1;
  }
};

File: leetcode-04/solutions/python3/0695.py
class Solution:
    def maxAreaOfIsland(self, grid: List[List[int]]) -> int:
        if not grid:
            return 0

        m = len(grid)
        n = len(grid[0])

        ans = 0
        self.visited = [[False for j in range(len(grid[0]))]
                        for i in range(len(grid))]

        for i in range(m):
            for j in range(n):
                ans = max(ans, self.dfs(grid, i, j))

        return ans

    def dfs(self, grid: List[List[int]], i: int, j: int) -> int:
        if i < 0 or j < 0 or i >= len(grid) or j >= len(grid[0]) or self.visited[i][j] or grid[i][j] == 0:
            return 0

        self.visited[i][j] = True

        return self.dfs(grid, i + 1, j) + self.dfs(grid, i - 1, j) + \
               self.dfs(grid, i, j + 1) + self.dfs(grid, i, j - 1) + 1
